idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,226 @@
+            while (currentPsi != null) {
+                val matchedNode = superDeclarationToCurrentDeclaration[currentPsi]
+                if (matchedNode != null) {
+                    val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+                    matchedNode.prepend(newNode)
+                    superDeclarationToCurrentDeclaration[superPsi] = newNode
+                    continue@outer
+                }
+                currentPsi = currentPsi.nextSibling
+            }
+            val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+            superDeclarationToCurrentDeclaration[superPsi] = newNode
+            sentinelTailNode.prepend(newNode)
+        }
+
+        insertMembersAccordingToPreferredOrder(sentinelHeadNode.toListSkippingNulls(), classLeftBrace, editor, currentClass)
+    }
+
+    private fun insertMembersAccordingToPreferredOrder(
+        symbolsInPreferredOrder: List<MemberEntry>,
+        classLeftBrace: PsiElement?,
+        editor: Editor,
+        currentClass: KtClassOrObject
+    ) {
+        var firstAnchor: PsiElement? = null
+        if (symbolsInPreferredOrder.first() is MemberEntry.NewEntry) {
+            val firstExistingEntry = symbolsInPreferredOrder.firstIsInstanceOrNull<MemberEntry.ExistingEntry>()
+            if (firstExistingEntry != null) {
+                firstAnchor = firstExistingEntry.psi.prevSiblingOfSameType() ?: classLeftBrace
+            }
+        }
+
+        val entryBatchAndAnchor = mutableListOf<Pair<List<KtCallableDeclaration>, PsiElement?>>()
+        var currentAnchor = firstAnchor
+        val currentBatch = mutableListOf<KtCallableDeclaration>()
+        fun updateBatch() {
+            if (currentBatch.isNotEmpty()) {
+                entryBatchAndAnchor += currentBatch.toList() to currentAnchor
+                currentBatch.clear()
+            }
+        }
+        for (entry in symbolsInPreferredOrder) {
+            when (entry) {
+                is MemberEntry.ExistingEntry -> {
+                    updateBatch()
+                    currentAnchor = entry.psi
+                }
+                is MemberEntry.NewEntry -> {
+                    currentBatch += entry.psi
+                }
+            }
+        }
+        updateBatch()
+
+        var firstElement: PsiElement? = null
+        for ((newDeclarations, anchor) in entryBatchAndAnchor) {
+            // TODO: implement proper referenceShortener
+            val inserted = insertMembersAfter(editor, currentClass, newDeclarations, anchor = anchor, referenceShortener = {})
+            if (firstElement == null) {
+                firstElement = inserted.first()
+            }
+        }
+
+        firstElement?.let {
+            moveCaretIntoGeneratedElement(editor, it)
+        }
+    }
+
+    private class DoublyLinkedNode<T>(val t: T? = null, var prev: DoublyLinkedNode<T>? = null, var next: DoublyLinkedNode<T>? = null) {
```

Comment:
Why do we need custom linked list here? Why not using usual list or java.util.LinkedList?

Reply:
Here I need pointers that points inside the linked list node for efficient insertion. If I use LinkedList, then the only API exposed that can do this is the `ListIterator`. But unfortunately `ListIterator` does not allow concurrent modifications, which is needed here to effectilvely insert the generated members in the right place.